lib/configuration/variables/sources/merge.js

```diff
@@ -0,0 +1,37 @@
+'use strict';
+
+const ServerlessError = require('../../../serverless-error');
+const isPlainObject = require('type/plain-object/is');
+
+module.exports = {
+  resolve: ({ params }) => {
+    if (!params || params.length < 2) {
+      throw new ServerlessError('${merge(...)} needs a sequence of values to merge');
+    }
+
+    if (params.every((item) => Array.isArray(item))) {
+      const result = [];
+      for (const param of params) result.push(...param);
+      return { value: result };
+    }
+
+    if (params.every((item) => isPlainObject(item))) {
+      const keys = {};
+
+      params.forEach((item, index) => {
+        Object.keys(item).forEach((key) => {
+          if (keys[key] !== undefined) {
+            throw new ServerlessError(
+              `duplicate key \`${key}\` in \${merge(...)}[${index}]; first seen in \${merge(...)}[${keys[key]}]`
+            );
+          }
+          keys[key] = index;
+        });
+      });
+
+      return { value: Object.assign({}, ...params) };
+    }
+
+    throw new ServerlessError('${merge(...)} needs a list of arrays or objects to merge');
```

Comment:
Let's reword to `Invalid arguments for variable "merge" (expected exclusive list of arrays or plain objects)` and add code `INVALID_VARIABLE_MERGE_PARAMS`